arex-schedule-web-api/src/main/java/com/arextest/schedule/service/LocalReplayService.java

```diff
@@ -0,0 +1,387 @@
+      }
+      // other cases
+      List<ReplayActionCaseItem> caseItems = Collections.emptyList();
+      while (true) {
+        // checkpoint: before sending page of cases
+        ReplayActionCaseItem lastItem =
+            CollectionUtils.isNotEmpty(caseItems) ? caseItems.get(caseItems.size() - 1) : null;
+        caseItems = replayActionCaseItemRepository.waitingSendList(replayPlan.getId(),
+            CommonConstant.MAX_PAGE_SIZE,
+            executionContext.getContextCaseQuery(),
+            Optional.ofNullable(lastItem).map(ReplayActionCaseItem::getId).orElse(null));
+
+        if (CollectionUtils.isEmpty(caseItems)) {
+          break;
+        }
+        List<String> caseIdList = batchCaseIdsMap.getOrDefault(contextIdentifier,
+            new ArrayList<>());
+        caseItems.forEach(replayActionCaseItem -> {
+          planItemIds.add(replayActionCaseItem.getPlanItemId());
+          caseIdList.add(replayActionCaseItem.getId());
+        });
+        batchCaseIdsMap.put(contextIdentifier, caseIdList);
+      }
+    }
+    response.setPlanId(replayPlan.getId());
+    cacheReplayActionItem(replayPlan.getReplayActionItemList(), planItemIds);
+    return CommonResponse.successResponse("queryReplayCaseId success!", response);
+  }
+
+  public QueryReplaySenderParametersResponse queryReplaySenderParameters(
+      QueryReplaySenderParametersRequest request) {
+    QueryReplaySenderParametersResponse response = new QueryReplaySenderParametersResponse();
+
+    List<ReplayActionCaseItem> caseItemList = replayActionCaseItemRepository.batchQueryById(
+        request.getCaseIds());
+    Map<String, ReplayActionItemForCache> planItemMap = new HashMap<>();
+    Map<String, ReplaySenderParameters> replaySenderParametersMap = new HashMap<>();
+    for (ReplayActionCaseItem caseItem : caseItemList) {
+      String planItemId = caseItem.getPlanItemId();
+      ReplayActionItemForCache replayActionItemForCache = planItemMap.getOrDefault(planItemId,
+          loadReplayActionItemCache(planItemId));
+      if (replayActionItemForCache == null) {
+        LOGGER.error("loadReplayActionItemCache failed, planItemId:{}", planItemId);
+        continue;
+      }
+      if (!planItemMap.containsKey(planItemId)) {
+        planItemMap.put(planItemId, replayActionItemForCache);
+      }
+      replaySenderParametersMap.put(caseItem.getId(),
+          buildReplaySenderParameters(caseItem, replayActionItemForCache));
+    }
+    response.setReplaySenderParametersMap(replaySenderParametersMap);
+    return response;
+  }
+
+  public boolean preSend(preSendRequest request) {
+    ReplayPlan replayPlan = replayPlanRepository.query(request.getPlanId());
+    SendLimiter sendLimiter = new SendRedisLimiter(replayPlan, redisCacheProvider);
+    ReplayActionCaseItem caseItem = restoreCase(request.getCaseId(), null);
+    if (sendLimiter.failBreak()) {
+      replayCaseTransmitService.updateSendResult(caseItem, CaseSendStatusType.EXCEPTION_FAILED);
```

Comment:
if caseItem is null, The method "updateSendResult" will throw nullpointerexception